### PR TITLE
[DX-547] Add OAuth Device Authorization Grant login flow

### PR DESCRIPTION
**Pull request by umair-ably**
## Motivation

The CLI currently authenticates via manually copy-pasted access tokens — users have to go to the Ably dashboard, create a token, and paste it into the terminal. This is friction-heavy, error-prone, and doesn't support token refresh.

This PR replaces that flow with **OAuth 2.0 Device Authorization Grant** ([RFC 8628](https://datatracker.ietf.org/doc/html/rfc8628)). The user runs `ably login`, gets a URL + code, approves in their browser, and the CLI receives tokens automatically. No secrets are ever pasted into a terminal. Tokens are stored locally and refreshed transparently.

This is the CLI-side counterpart to [ably/website#7962](https://github.com/ably/website/pull/7962), which adds the OAuth provider to the website.

### Commit 1: Add OAuth Device Authorization Grant login flow
Core infrastructure — the building blocks everything else depends on.

- **`src/services/oauth-client.ts`** — OAuth client implementing the Device Authorization Grant. Handles device code requests, token polling with exponential backoff, token refresh, and token revocation. Configurable via `OAuthConfig` with sensible defaults pointing to `ably.com`.
- **`src/services/token-refresh-middleware.ts`** — Transparent token refresh middleware. Wraps Control API calls and automatically refreshes expired access tokens using the stored refresh token before retrying.
- **`src/services/config-manager.ts`** — Extended to store/retrieve OAuth tokens (access token, refresh token, expiry) alongside existing account config. New methods: `getOAuthTokens`, `setOAuthTokens`, `clearOAuthTokens`.
- **`src/services/control-api.ts`** — New `getMe()` and `getAccounts()` methods for fetching user/account info after OAuth login.
- **`src/control-base-command.ts`** — Integrates the token refresh middleware so all Control API commands transparently refresh expired tokens.
- **`src/base-command.ts`** — Minor update to support the new auth flow.

### Commit 2: Add OAuth login and logout commands
The user-facing commands that use the infrastructure from commit 1.

- **`src/commands/accounts/login.ts`** — Rewritten to use Device Authorization Grant. Shows a verification URL + user code, opens a browser (unless `--no-browser`), polls for approval, then fetches account info and stores credentials. Supports `--alias` for named accounts. Multi-account: if the user belongs to multiple Ably accounts, prompts them to choose which to log in as.
- **`src/commands/accounts/logout.ts`** — Enhanced to revoke OAuth tokens on the server before clearing local credentials.
- **`src/commands/login.ts`** — Alias updated to match.

### Commit 3: Add multi-account selection and switching
UX for users who belong to multiple Ably organizations/accounts.

- **`src/commands/accounts/switch.ts`** — Expanded with interactive account selection. When switching, shows available accounts with their current app context. Supports switching by alias argument or interactive prompt.
- **`src/services/interactive-helper.ts`** — New interactive selection helpers used by login and switch flows.
- **`src/utils/slugify.ts`** — Utility to generate account aliases from account names (e.g., "My Company" → "my-company").

### Commit 4: Add tests for OAuth login flow
Comprehensive unit tests for all new code.

- **`test/unit/services/oauth-client.test.ts`** — Device code request, token polling (success, pending, slow-down, expiry, errors), token refresh, token revocation, configuration.
- **`test/unit/services/token-refresh-middleware.test.ts`** — Transparent refresh on 401, refresh failure handling, non-401 passthrough.
- **`test/unit/commands/accounts/login.test.ts`** — OAuth login flow, multi-account selection, alias handling, error cases, JSON output.
- **`test/unit/commands/accounts/logout.test.ts`** — Token revocation on logout, graceful handling when revocation fails.
- **`test/unit/commands/accounts/switch.test.ts`** — Account switching by alias.
- **`test/unit/utils/slugify.test.ts`** — Slug generation edge cases.
- **`test/helpers/mock-config-manager.ts`** — Extended with OAuth token mock methods.

## Review guidance

**Commit-by-commit review recommended** — each commit is self-contained and builds on the previous.

| Commit | What to focus on |
|--------|-----------------|
| 1 (infrastructure) | Security of token storage, correctness of the polling/backoff logic in `oauth-client.ts`, token refresh middleware edge cases |
| 2 (commands) | UX flow — especially the multi-account prompt, error messages, and JSON output structure |
| 3 (multi-account) | Interactive selection UX, alias generation, switch command behavior |
| 4 (tests) | Coverage completeness, mock fidelity |

### Key design decisions
- **Device Authorization Grant** (not Authorization Code + PKCE) — no localhost callback server needed, works in SSH sessions and containers, simpler UX
- **Token refresh is transparent** — commands don't need to know about OAuth; the middleware handles it in `ControlBaseCommand`
- **Multi-account support from day one** — users at agencies or with personal + work accounts can switch without re-authenticating
- **`--no-browser` flag** — for headless environments where `open` doesn't work; user manually copies the URL

## How to manually test

1. Ensure the website OAuth provider is running ([ably/website#7962](https://github.com/ably/website/pull/7962))
2. `ably login` — should open browser with device code page
3. Approve in browser → CLI should detect approval and store credentials
4. `ably apps list` — should work with the OAuth token
5. Wait for token expiry → next command should transparently refresh
6. `ably accounts switch` — should show interactive account picker
7. `ably accounts logout` — should revoke token and clear local credentials

DEMO VID

https://github.com/user-attachments/assets/004d3b7c-8bd1-4a8c-9ff6-016a523f0842

### CODE REVIEW

**Comment by umair-ably** [comment]
> @umair-ably this looks great, but OOI, did you consider [OAuth 2.0 Device Authorization Grant](https://datatracker.ietf.org/doc/html/rfc8628)?

@lmars I didn't initially but looking into it now... I'm on the fence...

Granted, it is _the_ solution for a headless environment but can you foresee substantial usage with that scenario? I'd expect devs are using the CLI locally, in which a 1-click authorise button in a browser is a significantly better experience than needing to enter an auth code yourself imo (which is also what we wanted to solve by getting rid of the need to copy the Control API Access Token). 

Unless we already know there's headless usage of the CLI (or foresee there will be a significant amount), I suggest we stick with this. We can quite easily introduce an additional login mode e.g. `ably login --use-device-code` (or check if the user can open a browser and then switch between the 2 login modes). 

WDYT?

**Comment by umair-ably** [comment]
@lmars updated 🙂 added a new vid in the PR description

File: src/commands/accounts/login.ts [hunk 1]
```diff
@@ -61,6 +65,10 @@ export default class AccountsLogin extends ControlBaseCommand {
       char: "a",
       description: "Alias for this account (default account if not specified)",
     }),
+    legacy: Flags.boolean({
```
[the remark this answers is not in the document]
Comment (umair-ably) [reply src/commands/accounts/login.ts]: Claude always thinking we need backwards compatibility... removed!

**Comment by umair-ably** [comment]
@mattheworiordan 

Logging in when a user has multiple accounts 

https://github.com/user-attachments/assets/ceceb664-e782-4cd8-bb6d-c7673f3d3555

File: src/commands/accounts/login.ts [hunk 2]
```diff
@@ -380,15 +315,98 @@ export default class AccountsLogin extends ControlBaseCommand {
     }
   }
 
-  private async openBrowser(url: string): Promise<void> {
+  private async oauthLogin(flags: BaseFlags): Promise<OAuthTokens> {
+    const oauthClient = new OAuthClient({
+      controlHost: flags["control-host"],
```
[the remark this answers is not in the document]
Comment (umair-ably) [reply src/commands/accounts/login.ts]: Addressed in e3bb6748 — the OAuth host now falls back to the stored `controlHost` from account config when the flag isn't explicitly passed. The naming (`controlHost` vs `dashboardHost`) is intentional: in our setup, the OAuth provider lives on the same host as the control/dashboard endpoints, and `--control-host` is the existing override mechanism for pointing to different environments. Renaming would be a larger cross-cutting change for minimal benefit.

File: src/commands/accounts/logout.ts [hunk 3]
```diff
@@ -88,6 +89,21 @@ export default class AccountsLogout extends ControlBaseCommand {
...
+      const oauthTokens = this.configManager.getOAuthTokens(targetAlias);
+      if (oauthTokens) {
+        const oauthClient = new OAuthClient({
+          controlHost: flags["control-host"],
+        });
+        // Best-effort revocation -- don't block on failure
+        await Promise.all([
+          oauthClient.revokeToken(oauthTokens.accessToken),
+          oauthClient.revokeToken(oauthTokens.refreshToken),
+        ]).catch(() => {});
```
[the remark this answers is not in the document]
Comment (umair-ably) [reply src/commands/accounts/logout.ts]: Addressed in e3bb6748 — logout now uses `flags["control-host"] || targetAccount?.controlHost` as fallback, so it picks up the stored host from the account config when the flag isn't passed. See line 82 in the current code.

File: src/control-base-command.ts [hunk 4]
```diff
@@ -26,6 +29,17 @@ export abstract class ControlBaseCommand extends AblyBaseCommand {
...
+
+      // Set up token refresh middleware for OAuth accounts
+      if (this.configManager.getAuthMethod() === "oauth") {
+        const oauthClient = new OAuthClient({
+          controlHost: flags["control-host"],
+        });
+        tokenRefreshMiddleware = new TokenRefreshMiddleware(
+          this.configManager,
+          oauthClient,
+        );
```
[the remark this answers is not in the document]
Comment (umair-ably) [reply src/control-base-command.ts:51]: Addressed in e3bb6748 — the token refresh middleware now uses `flags["control-host"] || account.controlHost`, falling back to the stored host from when the user originally logged in. See line 35 in the current code.

File: src/services/oauth-client.ts [hunk 5]
```diff
@@ -0,0 +1,319 @@
...
+  authorizationEndpoint: string;
+  clientId: string;
+  revocationEndpoint: string;
+  scopes: string[];
+  tokenEndpoint: string;
+}
+
+export interface OAuthClientOptions {
+  controlHost?: string;
+}
```
[the remark this answers is not in the document]
Comment (umair-ably) [reply src/services/oauth-client.ts:45]: Acknowledged — the naming could be clearer, but in practice the OAuth provider runs on the same host as the dashboard/control surface (defaulting to `ably.com`), and `--control-host` is the existing flag for pointing to different environments. Renaming to `dashboardHost` would require changes across multiple files for minimal functional benefit. Keeping as-is for now.

File: src/commands/accounts/login.ts [hunk 6]
```diff
@@ -380,15 +315,98 @@ export default class AccountsLogin extends ControlBaseCommand {
...
+      ? undefined
+      : ora("Waiting for browser authentication...").start();
+
     try {
-      // Use the 'open' package for cross-platform browser opening
-      // This handles platform differences safely and avoids shell injection
-      await open(url);
+      const tokens = await oauthClient.login(async (url) => {
+        await openUrl(url, this);
+      });
```
[the remark this answers is not in the document]
Comment (umair-ably) [reply src/commands/accounts/login.ts]: Fixed — in the current code, JSON mode emits a structured `awaiting_authorization` event instead of human-readable text (line 339-348). The `openUrl` call is guarded by `!flags["no-browser"]` and its log output goes to stderr via `this.logToStderr`, so it doesn't pollute JSON stdout. The spinner is also skipped in JSON mode (line 367-369).

File: src/commands/accounts/login.ts [hunk 7]
```diff
@@ -380,15 +315,98 @@ export default class AccountsLogin extends ControlBaseCommand {
...
+        ? `http://${host}/users/access_tokens`
+        : `https://${host}/users/access_tokens`;
+    }
+
+    if (!flags["no-browser"]) {
+      if (!this.shouldOutputJson(flags)) {
+        this.log("Opening browser to get an access token...");
+      }
+      await openUrl(obtainTokenPath, this);
+    } else if (!this.shouldOutputJson(flags)) {
```
[the remark this answers is not in the document]
Comment (umair-ably) [reply src/commands/accounts/login.ts]: No longer applicable — the legacy token login flow was completely removed in 3ebb6139. OAuth is now the only authentication method.

File: src/commands/accounts/login.ts [hunk 8]
```diff
@@ -76,87 +84,34 @@ export default class AccountsLogin extends ControlBaseCommand {
...
+            `OAuth login failed: ${error instanceof Error ? error.message : String(error)}`,
+          );
+          this.log("Falling back to manual token login...");
         }
-
-        await this.openBrowser(obtainTokenPath);
-      } else if (!this.shouldOutputJson(flags)) {
-        this.log(`Please visit ${obtainTokenPath} to create an access token`);
+        accessToken = await this.legacyTokenLogin(flags);
       }
```
[the remark this answers is not in the document]
Comment (umair-ably) [reply src/commands/accounts/login.ts]: No longer applicable — the legacy token login fallback was completely removed in 3ebb6139. There is no interactive fallback path anymore; OAuth is the only auth method.

File: src/commands/accounts/logout.ts [hunk 9]
```diff
@@ -88,6 +89,21 @@ export default class AccountsLogout extends ControlBaseCommand {
       }
     }
 
+    // Revoke OAuth tokens if this is an OAuth account
+    if (this.configManager.getAuthMethod(targetAlias) === "oauth") {
+      const oauthTokens = this.configManager.getOAuthTokens(targetAlias);
+      if (oauthTokens) {
+        const oauthClient = new OAuthClient({
+          controlHost: flags["control-host"],
+        });
```
[the remark this answers is not in the document]
Comment (umair-ably) [reply src/commands/accounts/logout.ts]: Addressed in e3bb6748 — logout now uses `flags["control-host"] || targetAccount?.controlHost` (line 82), falling back to the stored host from the account config when the flag isn't explicitly passed.

File: src/commands/accounts/logout.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (umair-ably) [reply src/commands/accounts/logout.ts]: Addressed — logout now wraps each revocation call in a 5-second timeout via `revokeWithTimeout` (lines 87-91), and the underlying `revokeToken` method also has a 10-second `AbortController` timeout (oauth-client.ts lines 180-188). Revocation failures are caught and logged without blocking logout (lines 92-102).

File: src/services/oauth-client.ts [hunk 10]
```diff
@@ -0,0 +1,319 @@
...
+   */
+  async refreshAccessToken(refreshToken: string): Promise<OAuthTokens> {
+    return this.postForTokens(
+      {
+        client_id: this.config.clientId,
+        grant_type: "refresh_token",
+        refresh_token: refreshToken,
+      },
+      "Token refresh",
+    );
```
[the remark this answers is not in the document]
Comment (umair-ably) [reply src/services/oauth-client.ts:233]: Addressed — `refreshAccessToken` now passes the current refresh token as `fallbackRefreshToken` to `postForTokens` (line 166), which passes it through to `parseTokenResponse` (line 237). If the server omits `refresh_token` in the response, the existing token is preserved (lines 246-249). This follows RFC 6749 §6 where the server may optionally issue a new refresh token.

File: src/services/oauth-client.ts [hunk 11]
```diff
@@ -0,0 +1,319 @@
...
+      client_id: this.config.clientId,
+      token,
+    });
+
+    try {
+      await fetch(this.config.revocationEndpoint, {
+        body: params.toString(),
+        headers: { "Content-Type": "application/x-www-form-urlencoded" },
+        method: "POST",
+      });
```
[the remark this answers is not in the document]
Comment (umair-ably) [reply src/services/oauth-client.ts]: Addressed — both `postForTokens` (15s timeout, lines 219-227) and `revokeToken` (10s timeout, lines 180-188) now use `AbortController` with `setTimeout` to prevent hanging. The timeout is cleared on completion to avoid leaks.

File: src/commands/accounts/login.ts [hunk 12]
```diff
@@ -168,15 +98,54 @@ export default class AccountsLogin extends ControlBaseCommand {
...
+      if (accounts.length === 1) {
+        selectedAccountInfo = accounts[0];
+      } else if (accounts.length > 1 && !this.shouldOutputJson(flags)) {
+        const picked =
+          await this.interactiveHelper.selectAccountFromApi(accounts);
+        selectedAccountInfo = picked ?? accounts[0];
+      } else {
+        // Multiple accounts in JSON mode or empty (backward compat: use first)
+        selectedAccountInfo = accounts[0];
+      }
```
[the remark this answers is not in the document]
Comment (umair-ably) [reply src/commands/accounts/login.ts:119]: Addressed — there's now an explicit `accounts.length === 0` guard (line 102-104) that falls back to the `/me` response when the accounts list is empty. The subsequent branches all use non-null assertions on `accounts[0]!` only when `accounts.length >= 1` is guaranteed by the conditional structure.

File: src/commands/accounts/login.ts [hunk 13]
```diff
@@ -168,15 +98,54 @@ export default class AccountsLogin extends ControlBaseCommand {
...
+      } else if (!alias) {
+        alias = slugifyAccountName(selectedAccountInfo.name);
+      }
+
+      // Store based on auth method
+      if (oauthTokens) {
+        this.configManager.storeOAuthTokens(alias, oauthTokens, {
+          accountId: selectedAccountInfo.id,
+          accountName: selectedAccountInfo.name,
+        });
```
[the remark this answers is not in the document]
Comment (umair-ably) [reply src/commands/accounts/login.ts]: Addressed — line 127 now spreads the OAuth tokens with the email from the `/me` API response: `{ ...oauthTokens, userEmail: user.email }`. This ensures the user email is always stored regardless of whether the token endpoint included it.

File: src/control-base-command.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (umair-ably) [reply src/control-base-command.ts:51]: Addressed in e3bb6748 — the token refresh middleware now uses `flags["control-host"] || account.controlHost` (line 35 of control-base-command.ts), falling back to the stored host from when the user originally logged in. The login command also persists `controlHost` to the account config (lines 134-137 of login.ts) so it's available for subsequent commands.

**Comment by umair-ably** [comment]
@claude review

**Comment by umair-ably** [comment]
@claude all feedback has been addressed. Review again

File: src/commands/accounts/login.ts [hunk 14]
```diff
@@ -351,62 +343,138 @@ export default class AccountsLogin extends ControlBaseCommand {
...
       }
     } catch (error) {
       this.fail(error, flags, "accountLogin");
     }
   }
 
-  private async openBrowser(url: string): Promise<void> {
+  private async oauthLogin(flags: BaseFlags): Promise<OAuthTokens> {
+    const oauthClient = new OAuthClient({
+      controlHost: flags["control-host"],
```
Comment (AndyTWF) [inline src/commands/accounts/login.ts]: I concur with Copilot on this one - technically the control host is `control.ably.net` whereas the oauth base url is `ably.com`. So they're not the same thing (both in terms of URL and what they are) and in theory don't need to be tied.

This flag is probably internal-only anyway, because nobody beyond us is going to be using anything but the prod control API. But for clear separation, I would rename the flag to `oauth-host` or similar to make it clear that it's not part of the control API.

File: src/services/oauth-client.ts [hunk 15]
```diff
@@ -0,0 +1,348 @@
...
+
+      if (error === "expired_token") {
+        throw new Error("Device code expired");
+      }
+
+      if (error === "access_denied") {
+        throw new Error("Authorization denied");
+      }
+
+      throw new Error(`Token polling failed: ${error}`);
```
Comment (AndyTWF) [inline src/services/oauth-client.ts]: This logs as `Error: Token polling failed: [object Object]` which isn't a great DX

File: src/services/oauth-client.ts [hunk 16]
```diff
@@ -0,0 +1,348 @@
...
+    interval: number,
+    expiresIn: number,
+  ): Promise<OAuthTokens> {
+    const deadline = Date.now() + expiresIn * 1000;
+    let currentInterval = interval;
+    let networkRetries = 0;
+    const maxNetworkRetries = 3;
+
+    while (Date.now() < deadline) {
+      await this.sleep(currentInterval * 1000);
```
Comment (AndyTWF) [inline src/services/oauth-client.ts]: Leave this going for long enough (~10s) and it'll hit a rate limit

File: src/services/oauth-client.ts [hunk 17]
```diff
@@ -0,0 +1,348 @@
...
+
+      if (response.ok) {
+        const data = (await response.json()) as Record<string, unknown>;
+        return this.parseTokenResponse(data);
+      }
+
+      let error: string;
+      try {
+        const errorData = (await response.json()) as Record<string, unknown>;
+        error = errorData.error as string;
```
Comment (AndyTWF) [inline src/services/oauth-client.ts]: In testing I believe that errorData.error is actually an object of status, message

Comment (AndyTWF) [reply src/services/oauth-client.ts]: Yeah so I've checked, in the website a 429 error takes the control API format which is nested. So something needs to be done to handle that (I think the website actually has a bug here in that it treats all requests that aren't control API requests as control API format).

In any case, we need an OAuth-compliant throttle response on this endpoint, which it looks like we anticipate below but website doesn't implement.

File: src/commands/accounts/login.ts (hunk 14 above)
Comment (AndyTWF) [reply src/commands/accounts/login.ts]: In fact, this definitely doesn't work. Because the step after OAuth is to go to the control API and get the accounts list. So because control API != oauth host, the flag can't be the same.

File: src/commands/accounts/logout.ts [hunk 18]
```diff
@@ -73,6 +74,52 @@ export default class AccountsLogout extends ControlBaseCommand {
...
+          // revokeToken() swallows its own errors and always resolves, so we
+          // don't need a .catch() here.
+          //
+          // We keep the timer handle and clear it once the race settles,
+          // otherwise a fast revocation leaves the timeout pending and the
+          // process hangs ~5s after printing "Successfully logged out".
+          const revokeWithTimeout = (token: string, timeoutMs = 5000) => {
+            let timer: NodeJS.Timeout | undefined;
+            const timeoutPromise = new Promise<void>((resolve) => {
+              timer = setTimeout(resolve, timeoutMs);
```
Comment (AndyTWF) [inline src/commands/accounts/logout.ts]: Should this promise reject when something goes wrong, so we can log that it timed out?

File: src/services/oauth-client.ts (hunk 16 above)
Comment (AndyTWF) [reply src/services/oauth-client.ts]: Should this use backoff and jitter instead of a flat "once a second" - would probably help with that rate limit

File: src/services/oauth-client.ts [hunk 19]
```diff
@@ -0,0 +1,348 @@
...
+
+  // --- Private helpers ---
+
+  private getOAuthConfig(
+    controlHost = DEFAULT_OAUTH_CONTROL_HOST,
+  ): OAuthConfig {
+    const host = controlHost;
+    const scheme = host.includes("local") ? "http" : "https";
+    return {
+      clientId: "gb-I8-bZRnXs-gF83jOWKQrUxPPWp_ldTfQtgGP0EFg",
```
Comment (AndyTWF) [inline src/services/oauth-client.ts:307]: I would add a note here that this is acceptable according to the RFC / the client id isnt secret

File: src/commands/accounts/logout.ts (hunk 18 above)
Comment (AndyTWF) [reply src/commands/accounts/logout.ts]: It also doesn't abort the fetch performed by revokeToken

File: src/commands/accounts/login.ts [hunk 20]
```diff
@@ -351,62 +343,138 @@ export default class AccountsLogin extends ControlBaseCommand {
...
-      // Use the 'open' package for cross-platform browser opening
-      // This handles platform differences safely and avoids shell injection
-      await open(url);
+      const tokens = await oauthClient.pollForToken(
+        deviceResponse.deviceCode,
+        deviceResponse.interval,
+        deviceResponse.expiresIn,
+      );
+
+      spinner?.succeed("Authentication successful!");
```
Comment (AndyTWF) [inline src/commands/accounts/login.ts]: I think in other places we end sentences with a full stop, so should here. Same on the one below.

**Review by AndyTWF — CHANGES_REQUESTED** [review]

File: src/commands/accounts/login.ts [hunk 21]
```diff
@@ -72,113 +69,95 @@ export default class AccountsLogin extends ControlBaseCommand {
...
-          );
-        }
+      const [{ account: meAccount, user }, accounts] = await Promise.all([
+        controlApi.getMe(),
+        controlApi.getAccounts(),
+      ]);
+
+      let selectedAccountInfo: { id: string; name: string };
+
+      if (accounts.length === 0) {
```
Comment (AndyTWF) [inline src/commands/accounts/login.ts:102]: In this and the === 1 case, i would explicitly tell the user what account they're working with before they start picking keys. That way there can be no confusion.

File: src/commands/accounts/login.ts [hunk 22]
```diff
@@ -72,113 +69,95 @@ export default class AccountsLogin extends ControlBaseCommand {
...
-
-        await this.openBrowser(obtainTokenPath);
-      } else if (!this.shouldOutputJson(flags)) {
-        this.log(`Please visit ${obtainTokenPath} to create an access token`);
-      }
-
-      accessToken = await this.promptForToken();
+    let oauthTokens: OAuthTokens;
+    try {
+      oauthTokens = await this.oauthLogin(flags);
```
Comment (AndyTWF) [inline src/commands/accounts/login.ts:81]: If you cancel at this point via CTRL + C you get something a bit unsightly:

```
Warning: Detected unsettled top-level await at Ably/cli/bin/run.js:37
await execute({
```

File: src/commands/accounts/switch.ts [hunk 23]
```diff
@@ -70,25 +72,201 @@ export default class AccountsSwitch extends ControlBaseCommand {
...
+        userEmail?: string;
+      };
+      alias: string;
+    }>,
+    flags: Record<string, unknown>,
+  ): Promise<boolean> {
+    const currentAlias = this.configManager.getCurrentAccountAlias();
+
+    // Try to fetch remote accounts using the current token
+    let remoteAccounts: AccountSummary[] = [];
```
Comment (AndyTWF) [inline src/commands/accounts/switch.ts:97]: On this command, if you don't specify any flags for control host, it will fail with "Access token may have expired or is invalid. The account was switched, but token verification failed."

File: src/commands/accounts/switch.ts
Comment (AndyTWF) [inline src/commands/accounts/switch.ts:49]: Flags need to be passed to this command?

File: src/utils/slugify.ts [hunk 24]
```diff
@@ -0,0 +1,52 @@
+/**
```
Comment (AndyTWF) [inline src/utils/slugify.ts:3]: There are libraries for creating slugs, can we use them so as not to re-invent the wheel?

https://www.npmjs.com/package/slugify

File: src/commands/login.ts [hunk 25]
```diff
@@ -3,14 +3,13 @@ import { Command } from "@oclif/core";
 import AccountsLogin from "./accounts/login.js";
 
 export default class Login extends Command {
```
Comment (AndyTWF) [inline src/commands/login.ts]: Should this extend ControlBaseCommand like the others so we get the lifecycle hooks?

File: src/commands/accounts/login.ts [hunk 26]
```diff
@@ -351,62 +352,154 @@ export default class AccountsLogin extends ControlBaseCommand {
...
+          userCode: deviceResponse.userCode,
+          verificationUri: deviceResponse.verificationUri,
+          verificationUriComplete: deviceResponse.verificationUriComplete,
+        },
+        flags,
+      );
+    } else {
+      this.log("");
+      this.log(
+        `  Your authorization code: ${chalk.bold.cyan(deviceResponse.userCode)}`,
```
Comment (AndyTWF) [inline src/commands/accounts/login.ts]: We have a helper I think for chalk?

File: src/commands/login.ts (hunk 25 above)
Comment (umair-ably) [reply src/commands/login.ts]: good spot - fixed! https://github.com/ably/ably-cli/pull/144/commits/c57e309738e1f3902781046fdaf344e85f882a1d

File: src/commands/accounts/login.ts (hunk 26 above)
Comment (umair-ably) [reply src/commands/accounts/login.ts]: hmm this was intentional... we have a resource helper which would make this cyan but not bold. 

Bold felt right for the auth code but probs is best to just use the resource format helper

Comment (umair-ably) [reply src/commands/accounts/login.ts]: https://github.com/ably/ably-cli/pull/144/commits/a842afff5e7e3908d6f6a9e61a5b87bdda2d814e

**Review by AndyTWF — APPROVED** [review]
